lib/sgtn-client/loader/single_loader.rb

```diff
@@ -12,34 +12,36 @@ module SingleLoader
       def load_bundle(component, locale)
         SgtnClient.logger.debug "[#{__FILE__}][#{__callee__}] component=#{component}, locale=#{locale}"
 
-        @single_bundle_loader ||= single_loader { |c,l| super(c,l) }
+        @single_bundle_loader ||= single_loader { |c, l| super(c, l) }
         id = CacheUtil.get_cachekey(component, locale)
         @single_bundle_loader.operate(id, component, locale)&.value
-      ensure
-        # delete thread from hash after finish
-        @single_bundle_loader.remove_object(id)
       end
 
       def available_bundles
         SgtnClient.logger.debug "[#{__FILE__}][#{__callee__}]"
 
         @single_available_bundles_loader ||= single_loader { super }
         @single_available_bundles_loader.operate(CONSTS::AVAILABLE_BUNDLES_KEY)&.value
-      ensure
-        @single_available_bundles_loader.remove_object(CONSTS::AVAILABLE_BUNDLES_KEY)
       end
 
       private
+
       def single_loader(&block)
+        loader = nil
         none_alive = proc { |_, thread| thread.nil? || thread.alive? == false }
         creator = proc do |id, _, *args|
           Thread.new do
             SgtnClient.logger.debug "start single loading #{id}"
-            block.call(*args)
+            begin
+              block.call(*args)
+            ensure
+              # delete thread from hash after finish
+              loader.remove_object(id)
```

Comment:
All the threads wait at line 17, and each of them will remove load threads from the hash.
This will cause the successive single loadings failed because load threads was removed.

Here the thread doesn't finish yet, but it will finish immediately.

Optionally, we can change this line 39 to make sure threads only are deleted after finish.
```
Thread.new do
  sleep 1
  loader.remove_object(id)
end
```

Reply:
> All the threads wait at line 17, and each of them will remove load threads from the hash. This will cause the successive single loadings failed because load threads was removed.
> 
> Here the thread doesn't finish yet, but it will finish immediately.
> 
> Optionally, we can change this line 39 to make sure threads only are deleted after finish.
> 
> ```
> Thread.new do
>   sleep 60
>   loader.remove_object(id)
> end
> ```

I understand now. What about just calling `@hash.delete(id)` before end of `operate` method?

Reply:
Hmmm, thinking about the original code again. I don't see why successive calls would fail if `operate` and  `remove_object` are using `@lock.synchronize`? One thread would release the lock first, before the other could acquire. If thread was removed from hash by a previous thread, the current will add it back to hash in `operate` right?